test/test-url.cpp

```diff
@@ -1154,3 +1154,34 @@ TEST(PunycodeTest, Safe)
     std::string encoded("http://safe.segments/");
     EXPECT_EQ(encoded, Url::Url(unencoded).punycode().str());
 }
+
+TEST(PunycodeTest, EmptySegment)
+{
+    ASSERT_THROW(Url::Url("http://foo..com/").punycode(), std::invalid_argument);
+}
+
+TEST(PunycodeTest, TrailingEmptySegment)
+{
+    ASSERT_THROW(Url::Url("http://foo../").punycode(), std::invalid_argument);
```

Comment:
I'm curious if `http://./` is treated as a host or a path. I'm not sure if that case will confuse your `check_hostname` function or not.